docs/development/design/build-images.rst

```diff
@@ -191,94 +299,122 @@ Examples of these versions are:
 The new image *may have different package versions* since there may be updates on the Ubuntu repositories.
-This conveys some small risk here, but in general we shouldn't require to add packages to the base images.
+This conveys some risk here, but in general we shouldn't require to add packages to the base images.
 
-Users with specific requirements could use ``build.system_packages`` and/or ``build.extras`` in the config file.
+In case we need an extra requirement for *all our images*,
+I'd recommend to add it when creating a new base image.
 
-How do we remove an old Python version?
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+If it's strongly needed and we can't wait for a new base image,
+we could install it at build time in a similar way as we do with ``build.apt_packages`` as a temporal workaround.
 
-At some point an old version of Python will be deprecated (eg. 3.4) and will be removed.
-To achieve this, we can just remove the Docker image affected: ``ubuntu20-py34``,
-once there are no users depending on it anymore.
+How do we create a mirror for each language?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A mirror can be created with ``wget`` together with ``rclone``:
+
+#. Download all the files from the official mirror:
+
+   .. code:: bash
+
+      # https://stackoverflow.com/questions/29802579/create-private-mirror-of-http-nodejs-org-dist
+      wget --mirror --convert-links --adjust-extension --page-requisites --no-parent -e robots=off http://nodejs.org/dist
+
+#. Upload all the files to S3:
+
+   .. code:: bash
+
+      # https://rclone.org/s3/
+      rclone sync -i nodejs.org s3:languages
+
+
+.. note::
+
+   Downloading a copy of the official mirror took 15m and 52Gb.
+
+How local development will work with the new approach?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Local development will require scripts to clone the official mirrors for each language and upload them to MinIO (S3).
+Besides, a script to define a set of Python version, pre-compile them and also upload them to S3.
 
-We will know which projects are using these images because they are pinning it in the config file.
-We could show a message in the build output page and also send them an email with the EOL date for this image.
 
 Deprecation plan
 ----------------
 
-It seems we have ~50Gb free on builders disks.
-Considering that the new images will be sized approximately (built locally as test):
-
-* ``ubuntu20-base``: ~5Gb
-* ``ubuntu20-py27``: ~150Mb
-* ``ubuntu20-py36``: ~210Mb
-* ``ubuntu20-py39``: ~20Mb
-* ``ubuntu20-conda47``: ~713Mb
+After this design document gets implemented and tested,
+all our current images (``stable``, ``latest``, ``testing``) will be deprecated and their usage will be de-motivated.
+However, we could keep them on our builders to give users a good time to migrate their projects to the new ones.
+New projects shouldn't be able to select these images and they will be forced to use ``build.os`` and ``build.languages``.
```

Comment:
This seems a bit overkill and user hostile. It will break the ability to import a new project using the same existing config a working project has, which I'm -1 on. I think we should add a new config version (2.1?), which is much more user-friendly.

Reply:
Won't supporting new projects to use the old images make us unable to move forward in the future? I feel that we are going to be blocked because of this if we accept new projects using an old/deprecated way to define builds.

Reply:
Sure, but we can achieve this with other methods, not making it break for users.

Reply:
The use case you want to support is "deleting an existing project using `build.image` and re-importing it should keep it working", right?

Do you have any idea about what other methods we can use here?

Reply:
I might suggest we detect when there are collisions: configs specifying both `build.image` and `build.os`/`build.languages`. We don't need to specify that new projects can *only* use the new config keys in that case, it's just the new default we prescribe everywhere.

I imagine we'll have the old build images around for a while.

Reply:
Yeah, I'm happy keeping the old build images for a while. However, I was just trying to "make the transition to the newer images faster" but maybe my approach was too aggressive 😄 . I will remove this line from the document.